README.md
## There's no place like ~/

These are my personal [dotfiles][]. They turn ordinary stock applications into
my own carefully crafted lightsaber.

    {///////\\\\\\\<#>}=============================================

Skeleton dotfile folder borrowed from [pengwynn][]. Heavily modified to fit my needs.

![screencap][]

## Get your own ##

If you're new to dotfiles, this probably isn't the best starter repository for
you. While these started as a clone of [Holman's][holman/dotfiles], many
utilities are unique to my workflow. I recommend other [dotfile frameworks][dotfiles]
out there.

### Installation ###

If you're crazy and want to clone my dots anyway:

- `git clone https://github.com/kingbin/dotfiles.git ~/.dotfiles`
- `cd ~/.dotfiles`
- `script/bootstrap`

The install script will symlink the appropriate files in `.dotfiles` to your
home directory.

Everything is configured and tweaked within `~/.dotfiles`.

All files and folders ending in `.symlink` or `.dotsymlink` get, you guessed it,
symlinked.
For example: `~/.dotfiles/bin.symlink` gets symlinked to `~/bin`
and `~/.dotfiles/gitconfig.dotsymlink` gets symlinked to. `~/.gitconfig`

## Main elements ##

There's a few special files in the hierarchy.

- **bin/**: Anything in `bin/` will get added to your `$PATH` and be made
  available everywhere.
- **topic/\*.zsh**: Any files ending in `.zsh` get loaded into your
  environment.
- **topic/\*.symlink**: Any files ending in `*.symlink` get symlinked into
  your `$HOME`. This is so you can keep all of those versioned in your dotfiles
  but still keep those autoloaded files in your home directory. These get
  symlinked in when you run `script/bootstrap`.
- **topic/\*.completion.sh**: Any files ending in `completion.sh` get loaded
  last so that they get loaded after we set up zsh autocomplete functions.

## Prior art ##

This project began as a fork of (and is heavily inspired by)
[Zach Holman's dotfiles][holman/dotfiles], whose topic-based symlinking
approach makes this so easy. I also have stolen freely from:

* [Mathias Bynens](http://github.com/mathiasbynens/dotfiles)
* [Yan Pritzker](http://github.com/skwp/dotfiles)
* [Josh Clayton](http://github.com/joshuaclayton/dotfiles)
* [Adam Jahnke](http://github.com/adamyonk/dotfiles)
* [Tom Ryder](http://github.com/tejr/dotfiles)
* [Steve Losh](http://github.com/sjl/dotfiles)

## ToDo ##

A good project is never truly done. I still plan to build out some additional brew installs I use.

## Share your dots

If you've got a great set of dots (or want to get started), check out
[dotfiles.github.com][dotfiles]. Ping @[octodots][] with great dot sets you've
found or tips and tricks for your favorite tools.

[dotfiles]: http://dotfiles.github.com
[pengwynn]: https://github.com/pengwynn/dotfiles
[octodots]: https://twitter.com/octodots
[holman/dotfiles]: https://github.com/holman/dotfiles
[screencap]: http://cl.ly/image/3k171S3E3k2q/Screen%20Shot%202013-02-07%20at%2011.06.41%20AM.png


CHANGELOG.md
### changes (HEAD -> master, origin/master)

>Author: Chris Blazek (chris.blazek@gmail.com)

>Author Date: Mon, 12 Mar 2018 16:25:29 -0500

>Commiter: Chris Blazek (chris.blazek@gmail.com)

>Commit Date: Mon, 12 Mar 2018 16:25:29 -0500

>Commit: 0303bca4407479e01c5fbb173b1d30802c2e791d 




### updating my dots

>Author: Chris Blazek (chris.blazek@gmail.com)

>Author Date: Mon, 12 Mar 2018 16:25:00 -0500

>Commiter: Chris Blazek (chris.blazek@gmail.com)

>Commit Date: Mon, 12 Mar 2018 16:25:00 -0500

>Commit: ebf626a82fef19057732a6649b944ded603cec78 




### spring cleaning

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Fri, 21 Apr 2017 12:17:26 -0500

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Fri, 21 Apr 2017 12:17:26 -0500

>Commit: ef86c60cabcba6d5e60e8d74f92322646daa2566 




### updated vim.dotsymlink

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Fri, 21 Apr 2017 11:49:52 -0500

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Fri, 21 Apr 2017 11:49:52 -0500

>Commit: 30ce97147bae7203a20060b61e9d3fa1459a62fd 




### updating submods

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Fri, 21 Apr 2017 11:34:09 -0500

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Fri, 21 Apr 2017 11:34:09 -0500

>Commit: 760bda0c5c0d606ca0b229471cd2d434b5e9263a 




### updating from mbp

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Fri, 21 Apr 2017 11:32:20 -0500

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Fri, 21 Apr 2017 11:32:20 -0500

>Commit: 677a62b4f88631b35a171e421e506e0b87da5cff 




### missing gitconfig settings and vim ver bump

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 18 May 2016 15:38:58 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 18 May 2016 15:38:58 -0500

>Commit: f901025f5c9e21b0e57b13eead067a0b62349d47 




### missing gitconfig settings

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 18 May 2016 15:20:22 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 18 May 2016 15:20:22 -0500

>Commit: f8fdd09e99610dd0771b8cc777e6049c0a9b83bf 




### updated vim

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 29 Jul 2015 13:19:38 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 29 Jul 2015 13:19:38 -0500

>Commit: c4762e5fcc7ff347536c950ddca3be0110bb9883 




### adding a couple useful goodies

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 29 Jul 2015 10:05:11 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 29 Jul 2015 10:05:11 -0500

>Commit: fa2dd8965ab037ba252d7e91a81ecf60f9ff3303 




### everyone wants to stash

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Thu, 2 Jul 2015 11:48:02 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Thu, 2 Jul 2015 11:48:02 -0500

>Commit: b644a575428fececeb14b58dd93f7aaa7afb65d2 




### updated vim config

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Fri, 19 Jun 2015 15:01:04 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Fri, 19 Jun 2015 15:01:04 -0500

>Commit: 41decd6e229fdb59991fc57482d9d8ba33a2105c 




### one more git alias

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 22 Apr 2015 08:45:23 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 22 Apr 2015 08:45:23 -0500

>Commit: 546c90d4956a94a59cfb82a420f21a7c7d34a76f 




### updated gitconfig & vim link

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 22 Apr 2015 08:40:49 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 22 Apr 2015 08:41:30 -0500

>Commit: 2640ce75096b3062a965f85d493ce2b36f987b0c 




### syntastic updates

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Thu, 12 Mar 2015 11:45:00 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Thu, 12 Mar 2015 11:45:00 -0500

>Commit: 6c3c1d532525a9d0486aad209e81dc467f708d65 




### syntastic updates

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Thu, 12 Mar 2015 11:40:52 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Thu, 12 Mar 2015 11:40:52 -0500

>Commit: 3db050789b749c7a1f7646a3c2064e42916f195a 




### vim updates

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 11 Mar 2015 15:16:15 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 11 Mar 2015 15:16:15 -0500

>Commit: 632e22537cc29ba387b07c1773396ca2637a668e 




### base16 theme setup

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Wed, 11 Mar 2015 14:33:43 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Wed, 11 Mar 2015 14:33:43 -0500

>Commit: 0e156a9c9605a1e961d199364ab507085f6ca86f 




### cleanup

>Author: Chris Blazek (cblazek@armt.com)

>Author Date: Tue, 10 Mar 2015 16:57:34 -0500

>Commiter: Chris Blazek (cblazek@armt.com)

>Commit Date: Tue, 10 Mar 2015 16:57:34 -0500

>Commit: 3c918e0547f56ef1de31cfed22441703ba6fb715 




### useful tools & alias for git

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Mon, 5 Jan 2015 16:38:47 -0600

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Mon, 5 Jan 2015 16:38:47 -0600

>Commit: f9fc312880b7ec119470a1ce07e3eedb4a27f310 




### final touches for the night

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Tue, 30 Dec 2014 22:27:54 -0600

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Tue, 30 Dec 2014 22:27:54 -0600

>Commit: 6239ed409f8243ea26d951e46d44eb28bfb3d287 




### updates to vim

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Tue, 30 Dec 2014 22:17:53 -0600

>Commiter: kingbin (chris.blazek@gmail.com)

>Commit Date: Tue, 30 Dec 2014 22:17:53 -0600

>Commit: 3202d4620250132df4bb5984155a98c767287b4f 




### still working at it

>Author: kingbin (chris.blazek@gmail.com)

>Author Date: Tue, 30 Dec 2014 22:05:07 -0600

>Commiter: kingbin (chris.blazek@gmail.com)


bin.symlink/dlsite.pl
#!/usr/bin/perl
#
# curlmirror.pl
#
# Mirrors a web site by using curl to download each page.
# The result is stored in a directory named "dest" by default.
# Temporary files are stored in "/tmp".
#
# Author: Kjell.Ericson@haxx.se
#
# Limitations:
#   All links are right now based from the root, so there are a lot
#   of "../../" in pages.
#
# History:
#
# 1999-11-19 v0.9 - Kjell Ericson - First version
# 1999-11-22 v0.10 - Kjell Ericson - Added some more flags
# 1999-12-06 v0.11 - Kjell Ericson - Relative paths were not correctecd
# 1999-12-06 v1.0  - Kjell Ericson - Satisfied and updated to v1.0
# 1999-12-07 v1.1  - Kjell Ericson - Added "-p"
# 1999-12-08 v1.2  - Kjell Ericson - Added "-l" and "-c"
# 1999-12-13 v1.3  - Kjell Ericson - Added match for images in stylesheets
# 2000-08-07 v1.4  - Kjell Ericson - Handles both ' and " in links.
# 2000-08-15 v1.5  - Kjell Ericson - Added -I.
# 2000-08-16 v1.6  - Kjell Ericson - Added multiple -I and -B.
# 2002-01-23 v1.7  - Anthony Thyssen - Changed the destination filename
# 2002-07-14 v1.8  - Kjell Ericson - Corrected a temp-filename error
#
$max_deep=1000;
$max_size=2;

$dest_dir="dest";
$default_name="index.html";
$tmp="/tmp";
$filecounter=0;

# For faster handling we have this regex:
$nonhtmlfiles="jpg|gif|png|zip|doc|txt|pdf|exe|java|js|css";

$help=
    "Usage: curlmirror.pl [flags] [url]\n".
    "\n".
    "-a <args>  : Curl specific arguments\n".
    "-B <url>   : Only retrieve URL below this URL (default is [url]).\n".
    "-b <name>  : Pattern that will be stripped from filename.\n".
    "-c         : Ignore CGI's (i.e URL's with '?' in them) (default off).\n".
    "-d <number>: Depth to scan on (default unlimited).\n".
    "-f         : Flat directory structure is made (be careful).\n".
    "-F         : Flat directory structure but use path in filename.\n".
    "-i <name>  : Default name for unknown filenames (default is 'index.html').\n".
    "-I <regex> : Don't handle files matching this pattern (default is \"\".\n".
    "-l         : Only load HTML-pages - no images (default is to load all).\n".
    "-o <dir>   : Directory to output result in (default is 'dest').\n".
    "-s <number>: Max size in Mb of downloaded data (default 2 Mb)\n".
    "-p         : Always load images (default is not to).\n".
    "-t <dir>   : Temporary directory (default is '/tmp').\n".
    "-v         : Verbose output.\n".
    "\n".
    "Example:\n".
    "curlmirror.pl http://www.perl.com/\n".
    "\nAuthor: Kjell.Ericson\@haxx.se\n";

for ($i=0; $i<=$#ARGV; $i++) {
    $arg=$ARGV[$i];
    if ($arg =~ s/^-//) {
        if ($arg =~ m/\?/) {
            print $help;
            exit();
        }
        if ($arg =~ m/a/) {
            $curl_args=$ARGV[++$i];
        }
        if ($arg =~ m/B/) {
            $base=$ARGV[++$i];
            if ($base !~ m/(http:\/\/[^\/]*)/i) {
                print "***Malformed -B(ase)\n";
                die($help);
            }
            $base=~ s/([+*~^()\\])/\\$1/g; # escape chars
            push @basematch, $base;
        }
        if ($arg =~ m/I/) {
            push @ignorepatt, $ARGV[++$i];
        }
        if ($arg =~ m/b/) {
            $strip_from_file=$ARGV[++$i];
        }
        if ($arg =~ m/c/) {
            $ignore_cgi=1;
        }
        if ($arg =~ m/d/) {
            $max_deep=$ARGV[++$i];
        }
        if ($arg =~ m/o/) {
            $dest_dir=$ARGV[++$i];
            $dest_dir=~ s/\/$//g;
        }
        if ($arg =~ m/t/) {
            $tmp=$ARGV[++$i];
            $tmp=~ s/\/$//g;
        }
        if ($arg =~ m/s/) {
            $max_size=$ARGV[++$i];
        }
        if ($arg =~ m/i/) {
            $default_name=$ARGV[++$i];
        }
        if ($arg =~ m/l/) {
            $only_html=1;
        }
        if ($arg =~ m/v/) {
            $verbose=1;
        }
        if ($arg =~ m/p/) {
            $picture_load=1;
        }
        if ($arg =~ m/f/) {
            $flat=1;
        }
        if ($arg =~ m/F/) {
            $flat=2;
        }
    } else { #default
        $start=$arg;
    }
}

$curl="curl -s $curl_args ";

if ($base eq "") {
    if ($start !~ m/(http:\/\/.+\/)/i) {
        if ($start =~ m/(http:\/\/.+)/i) {
            $start.="/";
        } else {
            print "***Malformed start URL ($start)\n";
            die($help);
        }
    }
    $base=$start;
    $base=~ s/\/[^\/]+$/\//; # strip docname
    $base=~ s/([+*~^()\\])/\\$1/g; # escape chars
    $basematch[0]=$base;
}


$follow_link{"start"}=0;

$linktmp="[ \n\r]*=[ \r\n]*)([\"'][^\"']*[\"']|[^ )>]*)";
%follow=(
         "(<[^>]*a[^>]+href$linktmp", "link",
         "(<[^>]*area[^>]+href$linktmp", "link",
         "(<[^>]*frame[^>]+src$linktmp", "link",
         );
if ($only_html == 0) {
    %follow=(%follow,
             "(BODY[^>]*\{[^}>]*background-image:[^>}]*url[(])([^\}>\) ]+)", "img", # for stylesheets
             "(<[^>]*img[^>]+src$linktmp", "img",
             "(<[^>]*body[^>]+background$linktmp", "img",
             "(<[^>]*applet[^>]+archive$linktmp", "archive",
             "(<[^>]*td[^>]+background$linktmp", "img",
             "(<[^>]*tr[^>]+background$linktmp", "img",
             "(<[^>]*table[^>]+background$linktmp", "img",
             );
}

$deep=0;
$found=1;
while ($found && $deep<$max_deep) {
    $found=0;
    foreach $url (keys %follow_link) {
        $current_depth=$follow_link{$url};
#        print STDERR ">$url $current_depth\n";
        if ($current_depth == $deep && $current_depth>=0 &&
            $total_size<$max_size*1024*1024) {
            $found=1;
            $current_depth++;
            if ($url eq "start") {
                delete $follow_link{$url};
                $url=$start;
                $url="stdin" if ($url eq "");
                $start="";
            }
            $follow_link{$url}=-1;
            $stop=0;

            $status_code=0;
            $content_type="";
            $real_url=$url;
            $real_url=~s/#(.*)//; # strip bookmarks before loading
            if ( $url !~ m/[ \n\r]/) {
                $filecounter++;
                $this_file_name="$filecounter..$real_url";
                $this_file_name =~ s/%([a-fA-F0-9][a-fA-F0-9])/chr hex $1/eg;
                #$this_file_name=~ s/[^a-zA-Z0-9.]+/_/g;
                $this_file_name=~ s/[^\w\d]+/_/g;
                $content_type="";

                print STDERR "Get $deep:$url\n" if ($verbose);
                $head=`$curl -D - -o "$tmp/$this_file_name" "$real_url"`;
                $filenames{$real_url}=$this_file_name;
                if ($head =~ m/Location: *["]?(.*)["]?/i) {
                    $loc=$1;
                    $loc=~ s/[\r\n]//g;
                    $loc=merge_urls($real_url, $loc);
                    if (accept_url($loc) ||
                        ($picture_load && $linktype{$real_url} eq "img")) {
                        `rm "$tmp/$this_file_name"`;
                        delete $filenames{$real_url};
                        $real_url=$loc;
                        $url=$loc;
                        print STDERR "Reget $deep:$url\n" if ($verbose);
                        $head=`$curl -D - -o "$tmp/$this_file_name" "$real_url"`;
                        $filenames{$real_url}=$this_file_name;
                        $follow_link{$real_url}=-1;
                    }
                }
                $total_size+=-s "$tmp/$this_file_name";

                if ($head =~ m/^HTTP[^\n\r]* ([0-9]+) ([^\n\r]*)/s) {
                    $status_code=$1;#." ".$2;
                }
                if ($head =~ m/[\n\r]Content-Type:(.*)[\r\n]/si) {
                    $content_type=$1;
                }
                $linktype{$real_url}=$content_type;
                if ($content_type !~ m/html/i) {
                    if ($only_html) { # remove this file
                        $total_size-=-s "$tmp/$this_file_name";
                        `rm "$tmp/$this_file_name"`;
                        delete $filenames{$real_url};
                    }
                } else {
                    $text=`cat "$tmp/$this_file_name"`;
                    if ($current_depth<$max_deep) {
                        $linktype{$real_url}="html";
                        $text="" if ($url =~ m/\#/);
                        foreach $search (keys %follow) {
                            while ($text =~ s/$search//si) {
                                $link=$2;
                                $link=~ s/[\"\']//g;
                                $link=~ s/#.*//;
                                $newurl=merge_urls($url, $link);
                                if ($ignore_cgi==0 || $newurl !~ m/\?/) {
                                    if (accept_url($newurl) ||
                                        ($picture_load && $follow{$search} eq "img")) {
                                        if (!exists $follow_link{$newurl}) {
                                            if ($only_html == 0 ||
                                                $newurl !~ m/\.($nonhtmlfiles)$/i) {
                                                $follow_link{$newurl}=$current_depth;
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    $deep++;
}
print STDERR "Max size exceeded ($total_size bytes)!\n" if ($total_size>=$max_size*1024*1024);
print STDERR "Total size loaded:$total_size bytes\n" if ($verbose);

foreach $url (keys %filenames) {
    local $destname=$url;
    $destname=~ s/$basematch[0]//;
    local $destdir=$destname;

    $destdir="" if ($destdir !~ m/\//);
    $destdir =~ s/\/[^\/]*$/\//;

    $destname=~ s/^.*\///g;
    $destname=~s/#(.*)//;
    local $bookmark=$1;

    $destname=~ s/[^a-zA-Z0-9.]/_/g; # strip chars we don't want in a filename
    $destdir=~ s/$strip_from_file// if ($strip_from_file ne "");
    $destdir=~ s/^([^\/]+):\/\//$1_/;
    $destdir=~ s/[^a-zA-Z0-9.\/_]/_/g;
    $destdir=~ s/(^\/)|(\/$)//g; # strip trailing/leading slashes

    if ($flat) {
        if ($flat==2) {
            $destdir=~ s/[\/:]/_/g;
            $destdir.="_";
        } else {
            $destdir="";
        }
        `mkdir -p "$dest_dir"`;
    } else {
        local $tmp="$dest_dir/$destdir";
        $tmp=~ s/^\///g;
        `mkdir -p "$tmp"`;
        $destdir.="/" if ($destdir ne "");
    }
    $destname=$default_name if ($destname eq "");
    $destname=$destdir.$destname if ($destdir ne "");
    if (($linktype{$url} =~ m/html/i) && ($destname !~ m/\.[s]?htm/i)) {
        $destname.=".html";
    }
    $destfile{$url}=$destname;
}

foreach $url (keys %filenames) {
    $name=$filenames{$url};
    $destname=$destfile{$url};

    if ($linktype{$url} !~ m/html/) {
        `mv "$tmp/$name" "$dest_dir/$destname"`;
    } else {
        $text=`cat "$tmp/$name"`;
        foreach $search (keys %follow) {
            $text=~ s/$search/"$1\"".make_file_relative($url,merge_urls($url, $2))."\""/sgie;
        }
        if (open(OUT, ">$dest_dir/$destname")) {
            print OUT $text;
            close(OUT);
        } else {
            print STDERR "Couldn't save file '$dest_dir/$destname'\n";
        }
        `rm "$tmp/$name"`;
    }
}

# Input: Base-URL, MakeRelative-URL
#
# Function: Convert and return "MakeRelativ-URL" to be relative
# to "Base-URL".
#
sub make_file_relative
{
    local ($from, $to)=@_;
    local $result="";
    local $sourcename=$destfile{$from};
    local $destname;
    local $bookmark;


    if ($to=~ s/(\#.*)$//) { # extract bookmarks
        $bookmark=$1;
    }

    $destname=$destfile{$to};

    if ($destname eq "") {
        return $to.$bookmark
    }


    $sourcename="" if ($sourcename !~ m/\//);

    $sourcename=~ s/\/[^\/]*$/\//; #strip filename
    do {
        $sourcename=~ m/^([^\/]*\/)/;
        local $dir=$1;
        if ($dir ne "") {
            $dir=~ s/([*.\\\/\[\]()+|])/\\$1/g;
            if ($destname =~ s/^$dir//) {
                $sourcename=~ s/^$dir//;
            } else {
                $dir="";
            }
        }
    } while ($dir ne "");
    $sourcename=~ s/[^\/]+\//..\//g; # Relative it with some ../

    $result="$sourcename$destname";
    $result=~ s/^\///g;

    return $result.$bookmark;
}


# Function: If you are viewing location "$base" which is a full URL, and
# click on "$new" that can be full or relative - where do you get? That
# is what this function returns.
#
# Input: base-URL, new-URL (where to go)
# Returns: a full format new-URL (without bookmark)
#
sub merge_urls
{
    local ($org, $new)=@_;
    local $url, $new;

    $new =~ s/[\"\']//g;

    if ($new =~ m/.*:/) {
        $url=$new;
    } elsif ($new eq "") {
        $url=$org;
    } else {
        if ($org =~ m/(.*):\/\/([^\/]*)(.*)$/) {
            local $prot=$1;
            local $server=$2;
            local $pathanddoc=$3;
            local $path;
            local $doc=$3;
            if ($pathanddoc=~ m/^(.*)\/(.*)$/) {
                $path=$1;
                $doc=$2;
            }
            $doc=~s/#(.*)//;
            local $bookmark=$1;

            if ($new =~ m/^#/) {
                $url="$prot://$server$path/$doc$new";
            } elsif ($new =~ m/^\//) {
                $url="$prot://$server$new";
            } else {
                $url="$prot://$server$path/$new";
                while ($url =~ s/\/[^\/]*\/\.\.\//\//){}
                while ($url =~ s/\.\///){};
            }
        }

    }
    return $url;
}

sub accept_url
{
    local ($url)=@_;
    local $ret=0;
    print "test  $url\n";
    for (@basematch) {
        if ($url =~ m/$_/) {
            $ret=1;
        }
    }
    return 0 if ($ret == 0); # No basematch

    for (@ignorepatt) {
        if ($url =~ m/$_/) {
            return 0;
        }
    }
    print "match $url\n";
    return 1;
}
